KOGITO-8740: data-index documentation (#409)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/data-index/data-index-core-concepts.adoc

= Data Index Core Concepts
:compat-mode!:
// Metadata:
:description: Data Index Service to allow to index and query audit data in {product_name}
:keywords: workflow, serverless, data, dataindex, data-index, index, service
// External pages
:cloud_events_url: https://cloudevents.io/
:graphql_url: https://graphql.org
:vertx_url: https://vertx.io/
:infinispan_url: https://infinispan.org/
:mongo_url: https://www.mongodb.com/
:postgresql_url: https://www.postgresql.org/
:dev_services_url: https://quarkus.io/guides/dev-services

> :flyway_quarkus_url: https://es.quarkus.io/guides/flyway

Author: ups!

[#data-index-deployments]
== {data_index_ref} scenarios

{data_index_ref} is distributed in different ways to allow deployment it in different topologies, and depending on how the data is indexed
The following sections describe the different options of {data_index_ref} deployments.

=== {data_index_ref} as a standalone service

It can be deployed explicitly referencing the *image*, starting a separated service inside a container. See xref:data-index/data-index-service.adoc[{data_index_ref} standalone].

image::data-index/data-index-standalone-service.png[Image of data-index deployment an external service]

This type of deployment requires to choose the right image depending on the persistence, specify the database connection properties, and the event consumption configuration.

[#data-index-dev-service]
=== {data_index_ref} service as Quarkus Development service
It also can be deployed, transparently as a *Quarkus Development Service* when the Quarkus Dev mode is used in the {product_name} application.

> When you use the {product_name} Process Quarkus extension, a temporary {data_index_ref} Service is automatically provisioned while the Quarkus application is running in development mode. When you use one of the following Quarkus extensions, the Dev Service is set up for immediate use:

Author: Replaced by
"When you use the {product_name} Process Quarkus extension, a temporary {data_index_ref} Service is automatically provisioned while the Quarkus application is running in development mode and the Dev Service is set up for immediate use."
Maybe would be simpler change 'following' by 'referenced'? 
@radtriste let me know what do you think?

> When you use the {product_name} Process Quarkus extension, a temporary {data_index_ref} Service is automatically provisioned while the Quarkus application is running in development mode. When you use one of the following Quarkus extensions, the Dev Service is set up for immediate use:

Reviewer: That new sentence is clearer yes :)

* Configure the {product_name} application to use the smallrye-kafka connector and the expected topic.

.Example of configuration in {product_name} application `application.properties` file to communicate with Kafka
[source,properties]
----
mp.messaging.outgoing.kogito-processinstances-events.connector=smallrye-kafka
mp.messaging.outgoing.kogito-processinstances-events.topic=kogito-processinstances-events
mp.messaging.outgoing.kogito-processinstances-events.value.serializer=org.apache.kafka.common.serialization.StringSerializer
----

[NOTE]
====
*Job service* is configured to send the JobEvents to the kafka topic `kogito-jobs-events` to have them available for {data_index_ref} consumption.
====

=== {data_index_ref} Quarkus extension and Jobs embedded addon
When {data_index_ref} functionality is added as a Quarkus extension to the workflow, there is no event configuration.
In this case, the data indexation is done internally, and all interactions are through the {data_index_ref} Gateway API.

== Additional resources

> * xref:core/timeouts-support.adoc[Timeouts in {product_name}]

Author: @radtriste I agree, I added the reference to knative and kafka eventing plus the timeout-showcase-example

// Referenced documentation pages
:job_service_guide: xref:supporting-services/jobs-service.adoc
:path_resolution_url: https://es.quarkus.io/blog/path-resolution-in-quarkus/#defaults

//Common constants
:data_index_ref: Data Index
:workflow_instance: process instance
:workflow_instances: {workflow_instance}s

In {product_name} architecture there is a dedicated supporting service that stores the data related to the {workflow_instances} and their associated jobs called *{data_index_ref}* service.
This service also provides a GraphQL endpoint allowing users to query that data and perform operations, also known as mutations in GraphQL terms.

The data processed by the {data_index_ref} service is usually received via events. The events consumed can be generated by the {product_name} runtime application or the {job_service_guide}[Job service] itself.
This event communication can be configured in different ways as described in the <<data-index-service-communication, {data_index_ref} communication configuration>> section.

The {data_index_ref} service uses Apache Kafka or Knative messaging to consume link:{cloud_events_url}[CloudEvents] messages from {product_name} services.

> The event data is indexed and stored in the database for querying via GraphQL.. These events contain information about units of work executed for a workflow.

Author: ups!

[NOTE]
====
The  `quarkus.http.root-path' and `quarkus.http.non-application-root-path` belong to the workflow application where the {data_index_ref} extension has been added. {path_resolution_url}[Here] there are more details about those properties and their default values.
====

[#data-index-service-communication]
== {data_index_ref} Service communication configuration

In order to index the data, {data_index_ref} allows multiple connection options to be able to consume the information provided by the different workflow applications.

The final goal is to receive the application-generated data related to the {workflow_instances} and jobs. The information that comes inside events, is indexed and stored in the database allowing it to be consumed through the provided GraphQL endpoint.

=== Knative Eventing

In order to interact with the {data_index_ref} separated service, using Knative eventing system eventing:

* Add the {data_index_ref} service and deployment, defining the Database connection properties and setting the `KOGITO_DATA_INDEX_QUARKUS_PROFILE` to `http-events-support.
* Specify the {data_index_ref} triggers to filter the

> .Example `DataIndex` resource with triggers definition (requires Knative KogitoInfra to be available):

Author: Just removed that reference to the 'old' concept

  broker: default
  filter:
    attributes:
      type: JobEvent
  subscriber:
    ref:
      apiVersion: v1
      kind: Service
      name: data-index-service-postgresql
    uri: /jobs
----

* Configure the {product_name} application to use the K_SINK as outgoing connection url

.Example of configuration in {product_name} application `application.properties` file to communicate with Knative
[source,properties]
----
mp.messaging.outgoing.kogito-processinstances-events.connector=quarkus-http
mp.messaging.outgoing.kogito-processinstances-events.url=${K_SINK}
mp.messaging.outgoing.kogito-processinstances-events.method=POST
----

[NOTE]
====
*Job service* needs also to be configured to send the events to the Knative K_SINK to have them available for {data_index_ref} related triggers.
====

=== Kafka messaging

To configure the communication between the {data_index_ref} Service and the workflow application through the Kafka eventing system, you must provide a set of configurations.

* Add the {data_index_ref} service and deployment, defining the Database connection properties and setting the `KOGITO_DATA_INDEX_QUARKUS_PROFILE` to kafka-events-support` (this value is set by default).

> * Configure the {product_name} application to use the K_SINK as outgoing connection url

Author: ups, yes!
I replaced by  'Configure the {product_name} application to use the smallrye-kafka connector and the expected topic.'

image::data-index/data-index-addon.png[Image of data-index as a Quarkus Extension]

More details are available in the xref:data-index/data-index-quarkus-extension.adoc[{data_index_ref} Quarkus Extension] section.

[#data-index-storage]
== Audit Data storage

In order to store the indexed data, {data_index_ref} needs some specific tables to be created. {data_index_ref} is ready to use link:{flyway_quarkus_url}[Quarkus flyway] for that purpose.

It's necessary to activate the migrate-at-start option to migrate the {data_index_ref} schema automatically.

There are more explanations about flyway migrations related in xref:persistence/postgresql-flyway-migration.adoc[Migrating Your PostgreSQL Database with a Kogito upgrade] section

[#data-index-graphql]
== GraphQL endpoint provided by {data_index_ref}

{data_index_ref} provides Graphql endpoint that allows users to interact with the stored audit data.
For more information about GraphQL see {graphql_url}[GraphQL]

[#data-index-ext-queries]
=== GraphQL queries for {workflow_instances} and jobs

> Use the following GraphQL queries to retrieve data about {workflow_instance}s and jobs.

Author: @cristianonicolai replaced by "This guide provides as example, some GraphQL queries that allow to retrieve data about {workflow_instance}s and jobs.". Does it sound fine?

Retrieve data from {workflow_instances}::
+
--
You can retrieve data about a specified {workflow_instance} from your Workflow definition.

.Example query
[source]
----
{
  ProcessInstances {
    id
    processId
    state
    parentProcessInstanceId
    rootProcessId
    rootProcessInstanceId
    variables
    nodes {
      id
      name
      type
    }
  }
}
----
--

> Retrieve data from jobs::

Reviewer 2: ```suggestion
Retrieve data from jobs:
```

Author: I can't remove that, it's needed to keep the guide format.

.Example query to retrieve a {workflow_instance} source file content
[source]
----
{ProcessInstances(where: { id: {equal: "1017afb1-5749-440e-8b9b-6b876bb5894d"}}){
  source
}}
----

[NOTE]
====
The workflow instance source field only will be available when `kogito-addons-quarkus-source-files` dependency is added on {product_name} runtime service `pom.xml` file.
[source,xml]
----
    <dependency>
      <groupId>org.kie.kogito</groupId>
      <artifactId>kogito-addons-quarkus-source-files</artifactId>
    </dependency>
----
====
--

Reschedule a job::
+
--
Reschedules a job using the `id`. The job `id` and other information are passed in the `data` parameter.

.Example mutation to reschedule a job
[source]
----
mutation{
  JobReschedule(
    id: "9674e3ed-8c13-4c3f-870e-7465d9ca7ca7",
    data:"{\"expirationTime\": \"2033-06-27T04:35:54.631Z\",\"retries\": 2}")
}
----
--

Cancel a job::
+
--
Cancels a job using the `id` passed as a parameter.

.Example mutation to cancel a job
[source]
----
mutation{
  JobCancel(
    id: "9674e3ed-8c13-4c3f-870e-7465d9ca7ca7")
}
----
--

[#data-index-graphql-ui]

> === Data Index graphical UI

Reviewer 2: ```suggestion
=== Data Index GraphQL UI
```

Data Index GraphQL UI is provided to interact with GraphQL endpoint.

image::data-index/data-index-graphql-ui.png[Image of data-index GraphQL UI]

> When the data-index is deployed as a separated service, this UI will be available at `/graphiql/` endpoint (i.e: at http://localhost:8180/graphiql/)

Reviewer 2: ```suggestion
When the Data Index is deployed as a standalone service, this UI will be available at `/graphiql/` endpoint (i.e: at http://localhost:8180/graphiql/)
```

To have the GraphQL UI available when the {data_index_ref} extension is deployed the property `quarkus.kogito.data-index.graphql.ui.always-include` needs to be enabled.

It will be accessible at: <quarkus.http.root-path><quarkus.http.non-application-root-path>/graphql-ui/ (i.e: http://localhost:8080/q/graphql-ui/)

[NOTE]
====
The  `quarkus.http.root-path' and `quarkus.http.non-application-root-path` belong to the workflow application where the {data_index_ref} extension has been added. {path_resolution_url}[Here] there are more details about those properties and their default values.
====

[#data-index-service-communication]
== {data_index_ref} Service communication configuration

In order to index the audit data, {data_index_ref} allows multiple connection options to be able to consume the information provided by the different workflow applications.

> The final goal is to receive the data generated by the application related with the {workflow_instances} and jobs. The information that comes inside events, is indexed and stored in the database allowing it to be consumed through the provided graphql endpoint.

Reviewer 2: ```suggestion
The final goal is to receive the application-generated data related to the {workflow_instances} and jobs. The information that comes inside events, is indexed and stored in the database allowing it to be consumed through the provided GraphQL endpoint.
```

metadata:
  name: data-index-service-postgresql-processes-trigger
spec:
  broker: default
  filter:
    attributes:
      type: ProcessInstanceEvent
  subscriber:
    ref:
      apiVersion: v1
      kind: Service
      name: data-index-service-postgresql
    uri: /processes
---
apiVersion: eventing.knative.dev/v1
kind: Trigger
metadata:
  name: data-index-service-postgresql-jobs-trigger
spec:
  broker: default
  filter:
    attributes:
      type: JobEvent
  subscriber:
    ref:
      apiVersion: v1
      kind: Service
      name: data-index-service-postgresql
    uri: /jobs
----

* Configure the {product_name} application to use the K_SINK as outgoing connection url

.Example of configuration in {product_name} application `application.properties` file to communicate with knative
[source,properties]
----
mp.messaging.outgoing.kogito-processinstances-events.connector=quarkus-http
mp.messaging.outgoing.kogito-processinstances-events.url=${K_SINK}
mp.messaging.outgoing.kogito-processinstances-events.method=POST
----

[NOTE]
====
*Job service* needs also to be configured to send the events to the knative K_SINK to have them available for {data_index_ref} related triggers.
====

> === Kafka Eventing

Reviewer 2: ```suggestion
=== Kafka messaging
```

* Add the {data_index_ref} service and deployment, defining the Database connection properties and setting the `KOGITO_DATA_INDEX_QUARKUS_PROFILE` to kafka-events-support` (this value is set by default).

* Configure the {product_name} application to use the K_SINK as outgoing connection url

.Example of configuration in {product_name} application `application.properties` file to communicate with knative
[source,properties]
----
mp.messaging.outgoing.kogito-processinstances-events.connector=smallrye-kafka
mp.messaging.outgoing.kogito-processinstances-events.topic=kogito-processinstances-events
mp.messaging.outgoing.kogito-processinstances-events.value.serializer=org.apache.kafka.common.serialization.StringSerializer
----

[NOTE]
====
*Job service* is configured to send the JobEvents to the kafka topic `kogito-jobs-events` to have them available for {data_index_ref} consumption.
====

=== {data_index_ref} Quarkus extension and Jobs embedded addon

When {data_index_ref} functionality is added as a Quarkus extension to the {product_name} application, there is no event configuration.

> In this case, the audit data indexation is internally done, and also all interactions through the {data_index_ref} gatewayAPI.

Reviewer 2: ```suggestion
In this case, the audit data indexation is done internally, and all interactions are through the {data_index_ref} Gateway API.
```

image::data-index/data-index-dev-service.png[Image of data-index deployment an Quarkus Dev Service]

More details are provided in the xref:data-index/data-index-service.adoc#data-index-dev-service-details[{data_index_ref} as a Quarkus Development service] section.

The {product_name} Process Quarkus extension sets up your Quarkus application to automatically replicate any {product_name} messaging events related to {workflow_instances} or jobs into the provisioned Data Index instance.

For more information about Quarkus Dev Services, see link:{dev_services_url}[Dev Services guide].

=== {data_index_ref} service as Quarkus extension
It can be included as part of the same {product_name} application using the *{data_index_ref} extension*, through the provided addons.

This scenario is specific to add the {data_index_ref} data indexing features and the GraphQL endpoint exposure inside a workflow application.

The communication with the workflow where the extension is added, is something internal to the application, allowing to simplify the communication between services and avoiding extra configuration for that purpose.

> In this case, the indexation has some limitations regarding only is able to index data from the workflows deployed in the same application.

Reviewer: ```suggestion
In this case, the indexation has some limitations: it is only able to index data from the workflows deployed in the same application.
```
